### PR TITLE
Fix the drawing position of OutlineInputBorder

### LINKED ISSUE #150109: OutlineBorder path for RTL InputDecorator is offsetted by 2 pixels
### Steps to reproduce

1. Copy the given test into packages/flutter/test/material/input_decorator_test.dart
2. Run the test

### Expected results

The test succeeds

### Actual results

The test failed.

To make it pass uncomment the line which adds a slight horizontal offset:

```dart
      // labelBottomRightLocalToBorder += const Offset(2, 0);
```

### Code sample

<details open><summary>Code sample</summary>

```dart

    testWidgets('OutlineBorder starts at the right position when text direction is RTL', (WidgetTester tester) async {
      Widget buildFrame(TextDirection textDirection) {
        return MaterialApp(
          home: Scaffold(
            body: Container(
              padding: const EdgeInsets.all(16.0),
              alignment: Alignment.center,
              child: Directionality(
                textDirection: textDirection,
                child: RepaintBoundary(
                  child: InputDecorator(
                    isFocused: true,
                    isEmpty: true,
                    decoration: InputDecoration(
                      labelText: labelText,
                      border: OutlineInputBorder(
                        borderRadius: BorderRadius.circular(30.0),
                        gapPadding: 0.0,
                      ),
                    ),
                  ),
                ),
              ),
            ),
          ),
        );
      }

      await tester.pumpWidget(buildFrame(TextDirection.rtl));
      final RenderBox borderBox = InputDecorator.containerOf(tester.element(findBorderPainter()))!;
      // Convert label bottom right offset to border path coordinate system.
      Offset labelBottomRightLocalToBorder = borderBox.globalToLocal(getLabelRect(tester).bottomRight);
      // labelBottomRightLocalToBorder += const Offset(2, 0);

      expect(findBorderPainter(), paints
        ..save()
        ..path(
          // The label bottom right corner should not be part of the border.
          excludes: <Offset>[
            labelBottomRightLocalToBorder,
          ],
          // The points just after the label bottom right corner should be part of the border.
          includes: <Offset>[
            labelBottomRightLocalToBorder + const Offset(1, 0),
            labelBottomRightLocalToBorder + const Offset(1, 1),
          ],
        )
        ..restore(),
      );
    });

```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

image

Code sample used to create the screenshot:

```dart

import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Container(
          padding: const EdgeInsets.all(16.0),
          alignment: Alignment.center,
          child: Directionality(
            textDirection: TextDirection.rtl,
            child: RepaintBoundary(
              child: InputDecorator(
                isFocused: true,
                isEmpty: true,
                decoration: InputDecoration(
                  labelText: 'label',
                  border: OutlineInputBorder(
                    borderRadius: BorderRadius.circular(30.0),
                    gapPadding: 0.0,
                  ),
                ),
              ),
            ),
          ),
        ),
      ),
    );
  }
}

```

</details>

### Logs

<details open><summary>Logs</summary>

```console

══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following TestFailure was thrown running a test:
Expected: Object or closure painting:
          'save()'
          'a path with that contains [Offset(757.0, 7.0), Offset(757.0, 8.0)] and does not contain
[Offset(756.0, 7.0)]'
          'restore()'
  Actual: _DescendantWidgetFinder:<Found 1 widget with widget matching predicate descending from
widget with widget matching predicate: [
            CustomPaint(renderObject: RenderCustomPaint#7d329),
          ]>
   Which: threw the following exception:
          It called drawPath with a path that unexpectedly did not contain Offset(757.0, 7.0).
          #0      _PathPaintPredicate.verifyArguments
(package:flutter_test/src/mock_canvas.dart:1263:11)
          #1      _DrawCommandPaintPredicate.match (package:flutter_test/src/mock_canvas.dart:934:5)
          #2      _TestRecordingCanvasPatternMatcher._evaluatePredicates
(package:flutter_test/src/mock_canvas.dart:848:27)
          #3      _TestRecordingCanvasMatcher.matches
(package:flutter_test/src/mock_canvas.dart:554:16)
          #4      _expect (package:matcher/src/expect/expect.dart:138:30)
          #5      expect (package:matcher/src/expect/expect.dart:56:3)
          #6      expect (package:flutter_test/src/widget_tester.dart:474:18)
          #7      main.<anonymous closure>.<anonymous closure>
(file:///home/bruno/Nevercode/flutter/packages/flutter/test/material/input_decorator_test.dart:1748:7)
          <asynchronous suspension>
          #8      testWidgets.<anonymous closure>.<anonymous closure>
(package:flutter_test/src/widget_tester.dart:183:15)
          <asynchronous suspension>
          #9      TestWidgetsFlutterBinding._runTestBody
(package:flutter_test/src/binding.dart:1025:5)
          <asynchronous suspension>
          #10     StackZoneSpecification._registerCallback.<anonymous closure>
(package:stack_trace/src/stack_zone_specification.dart:114:42)
          <asynchronous suspension>
          The complete display list was:
            * save()
            * drawPath(Path, Paint(PaintingStyle.stroke 2.0 StrokeJoin.miter up to 0.0;
Color(0xff6750a4)))
            * restore()

When the exception was thrown, this was the stack:
#4      main.<anonymous closure>.<anonymous closure> (file:///home/bruno/Nevercode/flutter/packages/flutter/test/material/input_decorator_test.dart:1748:7)
<asynchronous suspension>
#5      testWidgets.<anonymous closure>.<anonymous closure> (package:flutter_test/src/widget_tester.dart:183:15)
<asynchronous suspension>
#6      TestWidgetsFlutterBinding._runTestBody (package:flutter_test/src/binding.dart:1025:5)
<asynchronous suspension>
<asynchronous suspension>
(elided one frame from package:stack_trace)

This was caught by the test expectation on the following line:
  file:///home/bruno/Nevercode/flutter/packages/flutter/test/material/input_decorator_test.dart line 1748
The test description was:
  OutlineBorder starts at the right position when text direction is RTL
════════════════════════════════════════════════════════════════════════════════════════════════════

```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[!] Flutter (Channel [user-branch], 3.22.0-18.0.pre.719, on Ubuntu 24.04 LTS 6.8.0-35-generic, locale fr_FR.UTF-8)
```

</details>

### PR DESCRIPTION
**Pull request by hgraceb**
Fixes #150109.

1. There are two golden file tests broken. I have read the document for the golden file test, but I'm not sure if I should do something else.
2. Modify the test because of drawing position changes.

## Changes

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(MaterialApp(
    debugShowCheckedModeBanner: false,
    home: Scaffold(
      body: Padding(
        padding: const EdgeInsets.all(16.0),
        child: Column(
          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
          children: [
            Directionality(
              textDirection: TextDirection.ltr,
              child: RepaintBoundary(
                child: InputDecorator(
                  isFocused: true,
                  isEmpty: true,
                  decoration: InputDecoration(
                    labelText: 'label',
                    border: OutlineInputBorder(
                      borderRadius: BorderRadius.circular(30.0),
                      gapPadding: 0.0,
                    ),
                  ),
                ),
              ),
            ),
            Directionality(
              textDirection: TextDirection.rtl,
              child: RepaintBoundary(
                child: InputDecorator(
                  isFocused: true,
                  isEmpty: true,
                  decoration: InputDecoration(
                    labelText: 'label',
                    border: OutlineInputBorder(
                      borderRadius: BorderRadius.circular(30.0),
                      gapPadding: 0.0,
                    ),
                  ),
                ),
              ),
            ),
          ],
        ),
      ),
    ),
  ));
}
```

| Before                                   | After                                                  |
| ---------------------------------------- | ------------------------------------------------------ |
| Before | After |

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter/lib/src/material/input_border.dart [hunk 1]
```diff
@@ -482,14 +482,14 @@ class OutlineInputBorder extends InputBorder {
     // Draw top border from gap end to top right corner and draw top right corner.
     const double trCornerArcStart = (3 * math.pi) / 2.0;
     const double trCornerArcSweep = cornerArcSweep;
-    if (start + extent < scaledRRect.width - scaledRRect.trRadiusX) {
-      path.moveTo(scaledRRect.left + start + extent, scaledRRect.top);
+    if (start + extent < width - scaledRRect.trRadiusX) {
+      path.moveTo(start + extent, scaledRRect.top);
```
Comment (hgraceb) [inline packages/flutter/lib/src/material/input_border.dart:486]: Because `scaledRRect.width` is based on stroke width, replace it with `outer.width`.

**Comment by bleroux** [comment]
Thanks for the PR :pray: 
It would be great, if you can add screenshots of before and after (probably based on the code sample from https://github.com/flutter/flutter/issues/150109), thanks!

**Review by justinmc — COMMENTED** [review]
Thanks for the PR and for including the screenshots!

Can you explain why this algorithm should use the outer border? I want to get this right but I'm confused on exactly how this should work. I understand that a thick border could overlap with label text and the current algorithm (before this PR) would not consider that.

File: packages/flutter/lib/src/material/input_border.dart [hunk 2]
```diff
@@ -428,7 +428,7 @@ class OutlineInputBorder extends InputBorder {
   @override
   bool get preferPaintInterior => true;
 
-  Path _gapBorderPath(Canvas canvas, RRect center, double start, double extent) {
+  Path _gapBorderPath(Canvas canvas, RRect center, double width, double start, double extent) {
```
[the remark this answers is not in the document]
Comment (hgraceb) [reply packages/flutter/lib/src/material/input_border.dart]: I changed it.

File: packages/flutter/lib/src/material/input_border.dart [hunk 3]
```diff
@@ -482,14 +482,14 @@ class OutlineInputBorder extends InputBorder {
     // Draw top border from gap end to top right corner and draw top right corner.
     const double trCornerArcStart = (3 * math.pi) / 2.0;
     const double trCornerArcSweep = cornerArcSweep;
-    if (start + extent < scaledRRect.width - scaledRRect.trRadiusX) {
-      path.moveTo(scaledRRect.left + start + extent, scaledRRect.top);
```
[the remark this answers is not in the document]
Comment (hgraceb) [reply packages/flutter/lib/src/material/input_border.dart:486]: This is just to ensure that the appearance of `TextDirection.ltr` is consistent with that of `TextDirection.rtl`. I have reverted the code here. Perhaps we should open another issue for further discussion.

**Comment by hgraceb** [comment]
> Can you explain why this algorithm should use the outer border? I want to get this right but I'm confused on exactly how this should work. I understand that a thick border could overlap with label text and the current algorithm (before this PR) would not consider that.

I previously mentioned that the original scaledRRect.width was based on stroke width. Let me elaborate on my thoughts in more detail.

First, I share your view that the previous algorithm did not take border overlap into account. However, this change is consistent with the previous implementation, the only difference is that it ensures the appearance of `TextDirection.ltr` and `TextDirection.rtl` is the same.

Specifically, the border for `TextDirection.ltr` used a fixed start position that does not change with stroke width, while the border for `TextDirection.rtl` does change with stroke width. Therefore, to maintain consistency, I believe we can only choose to modify the algorithm for either `TextDirection.ltr` or `TextDirection.rtl`.

I personally feel that `TextDirection.ltr` is more in line with expectations because, as mentioned earlier, the previous approach did not consider overlap situations. So I ultimately decided to change `TextDirection.rtl` to use the same algorithm as `TextDirection.ltr`, which is independent of stroke width. Of course, we can also calculate the width using `scaledRRect.left + scaledRRect.right`, but considering reliability and readability, I ultimately chose to accept parameters from the outside.

**Comment by justinmc** [comment]
Thanks for the detailed response and for helping me figure this out.

 > Specifically, the border for TextDirection.ltr used a fixed start position that does not change with stroke width, while the border for TextDirection.rtl does change with stroke width.

Can you explain how the border for TextDirection.rtl changes with stroke width?

At first glance I'm thinking that we should be using the center of the border because that must be how the border radius is measured (`r` in my drawing, `scaledRRect.trRadiusX` in the code).

<details open>

<summary>Drawing of field's top right corner</summary>



</details>

**Comment by hgraceb** [comment]
> Can you explain how the border for TextDirection.rtl changes with stroke width?

The "TextDirection.rtl does change with stroke width" I mentioned earlier refers to the angle, not the radius.

We can look at the specific data by running the example above with an outer width of `400.0`.

------

If the stroke width is set to the [default value of 2.0](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/input_decorator.dart#L4839), when `TextDirection.ltr`, the start is `12.0`.

```dart
// Top left corner
if (scaledRRect.tlRadius != Radius.zero) {
  final double tlCornerArcSweep = math.acos(clampDouble(1 - start / scaledRRect.tlRadiusX, 0.0, 1.0));
  path.addArc(tlCorner, math.pi, tlCornerArcSweep);
}
```

However, when TextDirection.rtl, dx becomes `10.0`, because `scaledRRect.width = 379.0 - 1.0`.

```dart
if (start + extent < scaledRRect.width) {
  final double dx = scaledRRect.width - (start + extent);
  final double sweep = math.asin(clampDouble(1 - dx / scaledRRect.trRadiusX, 0.0, 1.0));
  path.addArc(trCorner, trCornerArcStart + sweep, trCornerArcSweep - sweep);
}
```

------

If we change the stroke width to `4.0`, while `TextDirection.ltr`, the start remains `12.0`.

```dart
// Top left corner
if (scaledRRect.tlRadius != Radius.zero) {
  final double tlCornerArcSweep = math.acos(clampDouble(1 - start / scaledRRect.tlRadiusX, 0.0, 1.0));
  path.addArc(tlCorner, math.pi, tlCornerArcSweep);
}
```

But when `TextDirection.rtl`, `dx` changes to `8.0`, because `scaledRRect.width = 378.0 - 2.0`.

```dart
if (start + extent < scaledRRect.width) {
  final double dx = scaledRRect.width - (start + extent);
  final double sweep = math.asin(clampDouble(1 - dx / scaledRRect.trRadiusX, 0.0, 1.0));
  path.addArc(trCorner, trCornerArcStart + sweep, trCornerArcSweep - sweep);
}
```

**Comment by hgraceb** [comment]
> At first glance I'm thinking that we should be using the center of the border because that must be how the border radius is measured (`r` in my drawing, `scaledRRect.trRadiusX` in the code).

I don't know if this approach would be better, but it seems like it should be a separate topic for discussion, as it implies that the entire border may need to be recalculated and redrawn.

**Comment by justinmc** [comment]
Thanks for the details, can you tell me how you are changing the stroke width of your border? Then I'll try this out tomorrow.

**Comment by hgraceb** [comment]
> Thanks for the details, can you tell me how you are changing the stroke width of your border? Then I'll try this out tomorrow.

You can directly modify the corresponding default value in the example above: [default value of 2.0](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/input_decorator.dart#L4839). Of course, you can also try overriding it using the theme.

**Review by justinmc — APPROVED** [review]
LGTM :+1: 

Thanks for helping me understand this solution. So start and extent are based on the outer width (the bounding box). The center rect shrinks as stroke width increases because it must fit inside of its bounding box including its padding. In this image, the yellow line shows the top is the same for both, but the green lines show that the center of the border has moved downward when the width increased:

border_width_center

Given that, the positioning of the top right side gap should be based on the outer width as well, as you've done in this PR.

| Before  this PR | After this PR |
| --- | --- |
| border_width | border_width_fixed |

**Review by bleroux — APPROVED** [review]
LGTM!
I'm a huge fan of such drawing fixes. Thanks for the fix and the details explanation. :pray: 

Some internal Google tests are failing. @justinmc thanks in advance for looking at these failures.

**Comment by justinmc** [comment]
Sorry for the delay here, I've reviewed all of the Google test failures and they are all expected changes to the border. In fact, the difference in all of them is pretty much imperceptible anyway. I've approved them.

I also approved the golden changes, which were similar. It should go green any minute but let me know if not.

**Pull request merged**